File: internal/algorithms/redis/token_bucket.go
package redis

import (
	"context"
	"fmt"
	"time"

	"github.com/AliRizaAynaci/gorl/core"
	"github.com/redis/go-redis/v9"
)

type TokenBucketLimiter struct {
	limit    int
	window   time.Duration
	client   *redis.Client
	prefix   string
	failOpen bool // fail-open policy flag
}

var tokenBucketLua = `
	local key = KEYS[1]
	local now = tonumber(ARGV[1])
	local refill = tonumber(ARGV[2])
	local limit = tonumber(ARGV[3])
	local window = tonumber(ARGV[4])

	local bucket = redis.call("HMGET", key, "tokens", "last_refill")
	local tokens = tonumber(bucket[1]) or limit
	local last_refill = tonumber(bucket[2]) or now

	local elapsed = now - last_refill
	local rate = refill / window
	local new_tokens = math.min(limit, tokens + elapsed * rate)
	if new_tokens < 1 then
		redis.call("HMSET", key, "tokens", new_tokens, "last_refill", now)
		redis.call("EXPIRE", key, window)
		return 0
	else
		redis.call("HMSET", key, "tokens", new_tokens - 1, "last_refill", now)
		redis.call("EXPIRE", key, window)
		return 1
	end
`

func (r *TokenBucketLimiter) Allow(key string) (bool, error) {
	ctx := context.Background()
	now := time.Now().Unix()
	redisKey := fmt.Sprintf("%s:%s", r.prefix, key)
	result, err := r.client.Eval(ctx, tokenBucketLua, []string{redisKey},
		now, r.limit, r.limit, int(r.window.Seconds()),
	).Result()
	if err != nil {
		// If backend is unavailable, follow fail-open or fail-close policy and return a custom error.
		if r.failOpen {
			return true, core.ErrBackendUnavailable
		}
		return false, core.ErrBackendUnavailable
	}
	allowed, ok := result.(int64)
	return ok && allowed == 1, nil
}

func NewTokenBucketLimiter(cfg core.Config) core.Limiter {
	opt, _ := redis.ParseURL(cfg.RedisURL)
	client := redis.NewClient(opt)
	return &TokenBucketLimiter{
		limit:    cfg.Limit,
		window:   cfg.Window,
		client:   client,
		prefix:   "gorl:tb",
		failOpen: cfg.FailOpen,
	}
}

File: storage/redis/lua/token_bucket.lua
local limit = tonumber(ARGV[1])
local now_us = tonumber(ARGV[2])
local ttl_ms = tonumber(ARGV[3])
local time_per_token_us = tonumber(ARGV[4])

local tokens = tonumber(redis.call("GET", KEYS[1]) or "0")
local last_refill = tonumber(redis.call("GET", KEYS[2]) or "0")

if last_refill == 0 then
  tokens = limit
  last_refill = now_us
else
  local elapsed = now_us - last_refill
  local new_tokens = math.floor(elapsed / time_per_token_us)
  if new_tokens > 0 then
    tokens = tokens + new_tokens
    if tokens > limit then
      tokens = limit
    end
    last_refill = last_refill + (new_tokens * time_per_token_us)
  end
end

local allowed = 0
if tokens > 0 then
  allowed = 1
  tokens = tokens - 1
end

local elapsed_since_refill = now_us - last_refill
local next_token_delay = time_per_token_us - elapsed_since_refill
if next_token_delay < 0 then
  next_token_delay = 0
end

local missing_tokens = limit - tokens
local reset_us = 0
if missing_tokens > 0 then
  reset_us = (missing_tokens * time_per_token_us) - elapsed_since_refill
  if reset_us < 0 then
    reset_us = 0
  end
end

redis.call("SET", KEYS[1], tokens, "PX", ttl_ms)
redis.call("SET", KEYS[2], last_refill, "PX", ttl_ms)

local retry_after_us = 0
if allowed == 0 then
  retry_after_us = next_token_delay
end

return {allowed, tokens, reset_us, retry_after_us}
